### PR TITLE
Generating raw code-based coverage data

### PR DESCRIPTION
**Pull request by amaurremi**
This addresses #245. ~~It adds a `--coverage` option that generates a `.profraw` data file with source-based coverage information~~ It adds a `coverage` command that can be used to generate [source-based coverage information](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/source-based-code-coverage.html) on program inputs generated by the fuzzer. ~~However, the user still needs to manually merge/index the generated raw files into a `.profdata` file~~ It produces a `.profdata` file which can be used for detailed reports and visualization. This is described in the README.

Thanks to @abizjak for helping debug a linking issue due to incompatible LLVM coverage-instrumentation flags.

### CODE REVIEW

File: README.md
Comment (nagisa) [inline README.md:73]: This appears to be somewhat inaccurate. At least as per the test added below the coverage is saved to `coverage_output.prof` despite `--coverage coverage_output` being specified as the argument.

Seems to me that this detail is important enough that its worth finding a way to describe this behaviour precisely.

File: README.md
Comment (nagisa) [inline README.md:64]: This documentation is probably better suited for [the book](https://github.com/rust-fuzz/book)?

File: README.md
Comment (nagisa) [inline README.md:76]: This links to the section about installing the tools, rather than merging the coverage files (which is the next section).

File: README.md
Comment (nagisa) [inline README.md:103]: Maybe "at least LLVM 11" would be better? The tools shipped with LLVM 12 will most likely do just as well if not better than the LLVM11 ones.

File: README.md
Comment (nagisa) [inline README.md:106]: Is "this" in "ensure this" the successful termination? Would fuzzer finding a crash count as a successful termination?

File: src/project.rs
Comment (nagisa) [inline src/project.rs:171]: Is there a bug that can be referenced here? It seems fairly unfortunate to have to tone down sanitization when the coverage is enabled. There is no obvious indication that one should not enable the `--coverage` option, for instance, and ideally there wouldn't be any need for it.

File: src/project.rs
Comment (nagisa) [inline src/project.rs:250]: You can pass `Path`/`PathBuf` to `env` directly, avoiding the conversion back to `String` and thus this failure mode.

**Review by nagisa — COMMENTED** [review]
Pretty nice!

File: README.md
Comment (amaurremi) [reply README.md:64]: I completely agree; however, this would require starting a coverage-report-generation section in the book and it might take some time to do. I suggest to keep it here until one of us finds the time to do that?

File: src/project.rs
Comment (amaurremi) [reply src/project.rs]: I agree that removing those flags significantly affects the fuzzing, so I put them back. They do cause linking errors when fuzzing some larger programs. It's been difficult to minimize those examples into reportable test cases but I hope to find the time to try.

**Comment by mykter** [comment]
The approach often taken to understanding coverage of fuzzing runs is to record the coverage hit by running through all of the entries in the corpus directory, rather than whilst doing actual fuzzing. The corpus directory contains all of the test cases required to understand the total coverage of the fuzzing run so far.

What do you think about changing this from an option to run, to its own command? e.g. `cargo fuzz coverage <target> <optional corpus dir>`

By doing this, we don't impact normal fuzzing at all, it is faster to run, and easier to use (no question of when to stop).

File: README.md
Comment (fitzgen) [reply README.md:64]: FWIW, I'm fine having this here for now.

File: src/options.rs
Comment (fitzgen) [inline src/options.rs:120]: It would be great if the average user didn't have to think about what to name the profile data file. Is there an existing convention that we can use by default?

File: tests/tests/main.rs
Comment (fitzgen) [inline tests/tests/main.rs:275]: Can we also assert that the `coverage_output.profraw` file exists?

**Review by fitzgen — COMMENTED** [review]
Thanks for all this @amaurremi! This is a highly desired feature, and you'll make a lot of people happy by implementing this!

I also think that a new subcommand rather than adding a flag to the existing `cargo fuzz run` subcommand (i.e. `cargo fuzz coverage ...` rather than `cargo fuzz run --coverage ...`) is the way to go here.

File: src/options.rs
Comment (mykter) [reply src/options.rs:120]: clang uses `default.profraw` if not told otherwise:

> When the program exits it will write a raw profile to the path specified by the LLVM_PROFILE_FILE environment variable. If that variable does not exist, the profile is written to default.profraw in the current directory of the program. 

https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program

Comment (amaurremi) [reply src/options.rs]: I am now generating `default-<corpus file name>.profdata` files for each individual input file from the corpus, and merging them in the end using the profdata tool. The user doesn't need to provide names for the raw files.

File: tests/tests/main.rs
Comment (amaurremi) [reply tests/tests/main.rs]: Yes, that makes sense. I now added that check right after we generate that file (i.e. outside of the test).

Comment (amaurremi) [reply tests/tests/main.rs]: (And also to the test.)

File: README.md
Comment (mykter) [inline README.md:102]: Now the approach has changed we need a new step, running cargo fuzz coverage.

File: README.md
Comment (mykter) [inline README.md:86]: Are there any issues with running `fuzz coverage` whilst fuzzing is ongoing (e.g. in another terminal)?
Similarly with alternating between fuzz run and fuzz coverage?

If it potentially causes issues it would be helpful to document them (if we can't work around them with, say, separate target binaries for coverage-instrumented vs fuzzing-instrumented).

Comment (amaurremi) [reply README.md:86]: I tried running `cargo fuzz run` and `cargo fuzz coverage` on the same target (a WebAssembly interpreter) in parallel and didn't encounter issues. The `coverage` command generated information for the corpus data that was available at the moment it was run, and I was able to re-run it while the fuzzer was still running. I was also able to re-run the fuzzer while the coverage information was being produced. 
However, note that alternating between the `run` and `fuzz` commands involves rebuilding the target because they require different compiler options.
Are there more potential issues you would be looking for? I imagine that in general one would want to avoid running the two modes in parallel to maximize the computing resources available to the fuzzer.

Comment (mykter) [reply README.md:86]: > I was also able to re-run the fuzzer while the coverage information was being produced.

I don't fully understand how this works, but if it does then that's great!

My confusion is that there is only a single target binary path, so it seems we either have a binary with coverage instrumentation or not, and will never have both at the same time. In the scenario you tested, how do we know which version is going to get run? Might the coverage loop switch to the non-coverage-instrumented binary half way through, when it gets compiled for `cargo fuzz run`?

File: src/options.rs
Comment (fitzgen) [inline src/options.rs:119]: I think the `skip = false` bit isn't necessary, is it? By default structopt will not skip fields.

Or is this trying to make this option skipped, so that it is only exposed with `cargo fuzz coverage`?

File: src/options/coverage.rs
Comment (fitzgen) [inline src/options/coverage.rs:10]: Similarly, I think that non-`Option`/non-`Vec` fields are required by default, so this `required(true)` annotation shouldn't be necessary here.

File: src/project.rs
Comment (fitzgen) [inline src/project.rs:620]: Calling `.output()` and then not using the output is a bit inefficient. `.output()` will buffer up stdout and stderr and return them, but if we don't actually care about the output, and we also don't want it printed to *our* stderr/stdout, then we can do 

```rust
cmd
    .stdout(std::process::Stdio::null())
    .stderr(std::process::Stdio::null())
```

when constructing the command, and then run the command via

```rust
let status = cmd.status().with_context(|| ...)?;
anyhow::ensure!(status.success, "command exited with failure code {}: {:?}", status.code, cmd);
```

(Note: we also want to check that the command exited successfully, as the above snippet does)

**Review by fitzgen — COMMENTED** [review]
Nice! A couple nitpicks to fix before we merge this but, once fixed, I think this is good to go! Thanks again @amaurremi!

File: src/options.rs
Comment (amaurremi) [reply src/options.rs:119]: > Or is this trying to make this option skipped, so that it is only exposed with cargo fuzz coverage?

Yes, the syntax here is `skip = <default value>`, which says that the option should not be shown to the user and specifies a default value. In our case the default value is `false`. I agree that this looks confusing.

My intention was to avoid exposing this flag to the user. It's a build flag that should always be used with the `coverage` command, and the code ensures that this flag is set with that command.

File: src/options/coverage.rs
Comment (amaurremi) [reply src/options/coverage.rs:10]: Ah, OK. Could you help me understand how this case is different from similar instances in options for the other command? For example, [here](https://github.com/rust-fuzz/cargo-fuzz/blob/master/src/options/run.rs#L10).

Comment (fitzgen) [reply src/options/coverage.rs:10]: Unless I'm mistaken, those uses of `required(true)` can also be removed.

File: src/options.rs
Comment (fitzgen) [reply src/options.rs:119]: Ah okay, thanks for the explanation. Would you mind explaining this a bit more in the comment? Thanks.

File: README.md
Comment (amaurremi) [reply README.md:86]: That's a good point. I decided to just in case create a separate build target for `cargo fuzz coverage`.

File: src/project.rs
Comment (amaurremi) [reply src/project.rs]: I actually do want to preserve the output of those commands, especially if there is a failure. So I am just using `cmd.status...` similarly to what you suggested.

**Comment by amaurremi** [comment]
Thank you for all the feedback. Always happy to receive more suggestions if you have any.

File: src/options/tmin.rs
Comment (fitzgen) [inline src/options/tmin.rs:11]: Thanks for cleaning up all these other options!

**Review by fitzgen — APPROVED** [review]
Thanks for keeping with the pull request! Looks great, going to merge it now.

**Pull request merged**
